pkg/scheduler/framework/v1alpha1/registry_test.go

```diff
@@ -104,3 +104,163 @@ func loadConfig(data []byte) (*config.KubeSchedulerConfiguration, error) {
 
 	return configObj, nil
 }
+
+type mockNoopPlugin struct {}
+
+func (p *mockNoopPlugin) Name() string {
+	return "MockNoop"
+}
+
+func NewMockNoopPluginFactory() PluginFactory {
+	return func(_ *runtime.Unknown, _ FrameworkHandle) (Plugin, error) {
+		return &mockNoopPlugin{}, nil
+	}
+}
+
+// reflect.DeepEqual() doesn't seem to work when comparing these two maps. 
+// This function serves to compare the two registries for equality.
+func isRegistryEqual(registryX Registry, registryY Registry) bool {
```

Comment:
did you try the k8s fork?
https://github.com/kubernetes/kubernetes/blob/master/third_party/forked/golang/reflect/deep_equal.go#L234

Reply:
@neolit123 I tried to use your suggestion basically as follows with no luck:
```
import "k8s.io/kubernetes/third_party/forked/golang/reflect"

var equalities = reflect.Equalities{}
equalities.DeepEqual(scenario.expected, scenario.registry)
```
Then I looked and found another candidate which also didn't work:
```
// https://github.com/kubernetes/apimachinery/blob/master/pkg/api/equality/semantic.go
equality.Semantic.DeepEqual(scenario.expected, scenario.registry)
```

Reply:
I find it surprising that deepequal didn't work, can you try cmp? see example here: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithmprovider/registry_test.go#L105

Reply:
@ahg-g I agree since everything I can see in the map matches. I went and setup a test to use @neolit123's suggestion again. The code is as follows:
```
func TestMerge(t *testing.T) {
	tests := map[string]struct {
		primaryRegistry Registry
		registryToMerge Registry
		expected        Registry
		shouldError     bool
	}{
		"goodMerge": {
			primaryRegistry: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
			},
			registryToMerge: Registry{
				"pluginFactory2": NewMockNoopPluginFactory(),
			},
			expected: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
				"pluginFactory2": NewMockNoopPluginFactory(),
			},
			shouldError: false,
		},
		"badMergeDuplicateFactory": {
			primaryRegistry: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
			},
			registryToMerge: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
			},
			expected: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
			},
			shouldError: true,
		},
	}

	for scenarioName, scenario := range tests {
		err := scenario.primaryRegistry.Merge(scenario.registryToMerge)

		if err == nil && scenario.shouldError == true {
			t.Errorf("Test %s: Merge() Failed to return Error.", scenarioName)
			continue
		}

		var equalities = reflect.Equalities{}
		if !equalities.DeepEqual(scenario.expected, scenario.primaryRegistry) {
			t.Errorf("Test %s: Failed to Merge(). Expected %v. Got %v instead.", scenarioName, scenario.expected, scenario.primaryRegistry)
		}

		// if !isRegistryEqual(scenario.expected, scenario.primaryRegistry) {
		// 	t.Errorf("Test %s: Failed to Merge(). Expected %v. Got %v instead.", scenarioName, scenario.expected, scenario.primaryRegistry)
		// }
	}
}
```
The test fails and the output is:
```
API server listening at: 127.0.0.1:26934
--- FAIL: TestMerge (7.45s)
    registry_test.go:181: Test goodMerge: Failed to Merge(). Expected map[pluginFactory1:0x16c02d0 pluginFactory2:0x16c02d0]. Got map[pluginFactory1:0x16c02d0 pluginFactory2:0x16c02d0] instead.
    registry_test.go:181: Test badMergeDuplicateFactory: Failed to Merge(). Expected map[pluginFactory1:0x16c02d0]. Got map[pluginFactory1:0x16c02d0] instead.
FAIL
```
The same result and basically the same output occurs when I use the Semantic.DeepEqual(). 

@ahg-g I am working on updating the code to use cmp now per your suggestion so another comment will follow shortly if I find the same issue there.

Reply:
@ahg-g Looks like the same issue occurs with cmp. See following:
```
func TestRegister(t *testing.T) {
	tests := map[string]struct {
		registry          Registry
		nameToRegister    string
		factoryToRegister PluginFactory
		expected          Registry
		shouldError       bool
	}{
		"goodRegister": {
			registry:          Registry{},
			nameToRegister:    "pluginFactory1",
			factoryToRegister: NewMockNoopPluginFactory(),
			expected: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
			},
			shouldError: false,
		},
		"badRegisterDuplicatePluginFactory": {
			registry: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
			},
			nameToRegister:    "pluginFactory1",
			factoryToRegister: NewMockNoopPluginFactory(),
			expected: Registry{
				"pluginFactory1": NewMockNoopPluginFactory(),
			},
			shouldError: true,
		},
	}

	for scenarioName, scenario := range tests {
		err := scenario.registry.Register(scenario.nameToRegister, scenario.factoryToRegister)

		if err == nil && scenario.shouldError == true {
			t.Errorf("Test %s: Register() Failed to return Error.", scenarioName)
			continue
		}

		if diff := cmp.Diff(scenario.expected, scenario.registry); diff != "" {
			t.Errorf("Test %s: Failed to Register(). (-want, +got): %s", scenarioName, diff)
		}

		// if !isRegistryEqual(scenario.expected, scenario.registry) {
		// 	t.Errorf("Test %s: Failed to Register(). Expected %v. Got %v instead.", scenarioName, scenario.expected, scenario.registry)
		// }
	}
}
```
Produces output:
```
matt@localhost v1alpha1]$ go test -run TestRegister
--- FAIL: TestRegister (0.00s)
    registry_test.go:225: Test goodRegister: Failed to Register(). (-want, +got):   v1alpha1.Registry{
        - 	"pluginFactory1": ⟪0x112e600⟫,
        + 	"pluginFactory1": ⟪0x112e600⟫,
          }
    registry_test.go:225: Test badRegisterDuplicatePluginFactory: Failed to Register(). (-want, +got):   v1alpha1.Registry{
        - 	"pluginFactory1": ⟪0x112e600⟫,
        + 	"pluginFactory1": ⟪0x112e600⟫,
          }
FAIL
exit status 1
FAIL	k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1	0.007s
[matt@localhost v1alpha1]$ 
```